railroadmedia/musora-content-services
# TP-175 - Needs Access logic

[MWP PR](https://github.com/railroadmedia/musora-web-platform/pull/1447)


**Design:**
- add decorator equivalent to `fetchSanity` to allow caller to customize the return package.
- add `processNeedAccess` to add the 'needs_access' property to the results
- needs_access should be set to true if the user doesn't have permissions that match the permission of the content.

**Future Improvements for Performance**
- Flip the default value for `processNeedAccess` to false, and only set it to true if the UserIsBasic OR we are accessing content that needs to be locked (eg: playlists).
- potentially pass in the user permissions, instead of grabbing them from the BE. This requires a bigger refactor of how user gets passed or stored for different methods. I'd like to avoid passing them in explicitly on each call.
- Get permissions from the user store once it's available. Waiting on Rob for this framework, then needs to be extended to include permissions.

This could also be handled in the groq query if we dynamically add the field:
`"need_access" : "<logic here to check user permissions>"` but this would need to be tested for performance vs the solution provided here. It can't be statically defined as the user's permissions are calculated at runtime so adding it to contentTypeConfig.js is not possible with the current design.
 
**Testing**
- Tests pass (I had to add a dummy flag for processNeedsAccess, I'm not sure this is a good idea, but adding a test framework for railcontent is a bit overkill)
-- when the user store is in place, adding a test harness for it would be a relatively easy extension in the testsuite.
- Update DevEndpoint.vue with `fetchByRailContentIds([380094,402204]).then(a => console.log('check needAccess list', a)); fetchByRailContentId(402204).then(a => console.log('check needAccess single', a));`
- Update sanity.js -> getUserPermissions `return [91];`
- Verify calling /devendpoint, check the console for the object and verify `needs_access` is set to true for the items that don't have 91 in the `permission_id` property. 
- Log out and call /devendpoint again, all results should have `needs_access` set to true.
